### PR TITLE
Rebuild state

### PR DESCRIPTION
**Pull request by tiagolobocastro**
rebuild: split rebuild module

splits the rebuild module into impl and api
no functional changes
this should help contain the next set of coming changes

-----------------------------------------------------

rebuild: split the rebuild state logic

Fixes the rebuild state by introducing and current and a pending state
the pending state is set by the nexus whenever it wants to perform
an operation on the rebuild. The pending state is reconciled when the
job's runner adjusts to the pending state, eg running -> paused

The states are now fully validated by a state machine. No catch all
"_ => ()", was used so the compiler should complain if a new state is
added.

If a pending state is "pending" then a new operation is rejected until
the previous completes.
Internal operations are allowed to override the pending state

Started to make the trait a first class citizen

bonus: fully implement the child src removal by waiting on the rebuild
job to complete before creating a new job to replace the removed one.
This means we now enable the commented out test

### CODE REVIEW

File: mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs [hunk 1]
```diff
@@ -61,24 +61,39 @@ impl Nexus {
...
+        job.as_client().start().context(CreateRebuildError {
+            child: name.to_owned(),
+            name: self.name.clone(),
+        })
+    }
+
+    /// Terminates a rebuild in the background
+    /// used for shutdown operations and
+    /// unlike the client operation stop, this command does not fail
+    /// as it skips previous client operations
```
Comment (paulyoong) [inline mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs]: Is it accurate to say it skips previous client operations? They may already be in progress. Is it more of a case of overwriting the desired state.

File: mayastor/src/replicas/rebuild/rebuild_impl.rs [hunk 2]
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
...
+}
+
+impl Default for RebuildStates {
+    fn default() -> Self {
+        Self {
+            current: RebuildState::Init,
+            pending: None,
+        }
+    }
+}
```
Comment (paulyoong) [inline mayastor/src/replicas/rebuild/rebuild_impl.rs]: Given that these are being set to their default values, could we use Default::default()?

File: mayastor/src/replicas/rebuild/rebuild_impl.rs [hunk 3]
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
...
+
+                        // The rebuild is not running so we need to notify here
+                        self.notify();
+                        Ok(())
+                    }
+                }
+            }
+            RebuildOperation::Pause => match self.state() {
+                S::Stopped | S::Failed | S::Completed => Err(e),
+                S::Init | S::Running | S::Paused => {
```
Comment (paulyoong) [inline mayastor/src/replicas/rebuild/rebuild_impl.rs]: Should the paused state be split out on its own so it resembles the other cases i.e. "for idempotence sake" on line 584?

File: mayastor/src/replicas/rebuild/rebuild_impl.rs [hunk 4]
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
...
+                S::Running => Ok(()),
+                S::Paused => {
+                    self.states.set_pending(S::Running, false)?;
+                    self.schedule();
+                    Ok(())
+                }
+            },
+            RebuildOperation::Fail => match self.state() {
+                S::Init | S::Stopped | S::Paused | S::Completed => Err(e),
+                S::Running | S::Failed => {
```
Comment (paulyoong) [inline mayastor/src/replicas/rebuild/rebuild_impl.rs]: Again, should the failed state be split out on its own?

File: mayastor/src/replicas/rebuild/rebuild_impl.rs [hunk 5]
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
...
+                    self.states.set_pending(S::Failed, override_pending)?;
+                    Ok(())
+                }
+            },
+            RebuildOperation::Complete => match self.state() {
+                S::Init | S::Paused | S::Stopped | S::Failed | S::Completed => {
```
Comment (paulyoong) [inline mayastor/src/replicas/rebuild/rebuild_impl.rs:627]: If its already completed should it just return Ok?

**Review by paulyoong — CHANGES_REQUESTED** [review]

File: mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs (hunk 1 above)
Comment (tiagolobocastro) [reply mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs]: skips -> overrides ?

Comment (paulyoong) [reply mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs]: I think so

File: mayastor/src/replicas/rebuild/rebuild_impl.rs (hunk 2 above)
Comment (tiagolobocastro) [reply mayastor/src/replicas/rebuild/rebuild_impl.rs]: Done

File: mayastor/src/replicas/rebuild/rebuild_impl.rs (hunk 3 above)
Comment (tiagolobocastro) [reply mayastor/src/replicas/rebuild/rebuild_impl.rs]: Done

File: mayastor/src/replicas/rebuild/rebuild_impl.rs (hunk 4 above)
Comment (tiagolobocastro) [reply mayastor/src/replicas/rebuild/rebuild_impl.rs]: Done

File: mayastor/src/replicas/rebuild/rebuild_impl.rs (hunk 5 above)
Comment (tiagolobocastro) [reply mayastor/src/replicas/rebuild/rebuild_impl.rs:627]: hmm I was trying to signal that you shouldn't have called complete twice as it only completes when you "complete"

File: mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs (hunk 1 above)
Comment (tiagolobocastro) [reply mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs]: Done

File: mayastor/src/replicas/rebuild/rebuild_impl.rs (hunk 5 above)
Comment (paulyoong) [reply mayastor/src/replicas/rebuild/rebuild_impl.rs:627]: Ok

**Review by paulyoong — APPROVED** [review]

**Pull request closed without merging**